Mathlib/NumberTheory/Transcendental/AnalyticPart.lean

```diff
@@ -0,0 +1,296 @@
+        have := Complex.analyticOrderAt_deriv_of_pos (n:= r) hf hr.symm (by grind)
+        rw [this]
+        exact ENat.coe_ne_top (r - 1)
+    · refine fun ho ↦ ⟨fun k hk ↦ ?_, ?_⟩
+      · have : analyticOrderAt (deriv^[k] f) z₀ ≠ 0 := by
+          rw [(Complex.analyticOrderAt_iterated_deriv f hf k (n+1)
+            ho.symm (by grind) hk.le), @Nat.cast_ne_zero]
+          grind
+        rw [analyticOrderAt_ne_zero] at this
+        exact this.2
+      · have := Complex.analyticOrderAt_iterated_deriv f hf (n+1) (n+1)
+          ho.symm (by grind) (by grind)
+        simp only [Function.iterate_succ, Function.comp_apply, tsub_self,
+          CharP.cast_eq_zero] at this
+        rw [AnalyticAt.analyticOrderAt_eq_zero] at this
+        · assumption
+        · exact iterated_deriv hf (n + 1)
+
+lemma analyticOrderAt_eq_nat_imp_iteratedDeriv_eq_zero
+    z₀ (n : ℕ) (f : ℂ → ℂ) (hf : AnalyticAt ℂ f z₀) :
+  analyticOrderAt f z₀ = n → (∀ k < n, deriv^[k] f z₀ = 0) ∧ (deriv^[n] f z₀ ≠ 0) := fun h ↦
+  (analyticOrderAt_eq_nat_iff_iteratedDeriv_eq_zero z₀ n f hf (h.symm ▸ ENat.coe_ne_top n)).mpr h
+
+lemma le_analyticOrderAt_iff_iteratedDeriv_eq_zero (n : ℕ) z₀
+  (f : ℂ → ℂ) (hf : AnalyticAt ℂ f z₀) (ho : analyticOrderAt f z₀ ≠ ⊤) :
+   (∀ k < n, (deriv^[k] f) z₀ = 0) → n ≤ analyticOrderAt f z₀ := by
+    intros hkn
+    have notTop (m : ℕ∞) : m ≠ ⊤ → ∃ n : ℕ, m = n :=
+      (Option.ne_none_iff_exists'.mp ·)
+    obtain ⟨m, Hm⟩ := notTop (analyticOrderAt f z₀) ho
+    rw [Hm, ENat.coe_le_coe]
+    rw [← analyticOrderAt_eq_nat_iff_iteratedDeriv_eq_zero z₀ m f hf ho] at Hm
+    by_contra! h
+    exact Hm.2 (hkn m h)
+
+lemma hasFPowerSeriesWithinAt_nhds_iff (f : ℂ → ℂ) (p : FormalMultilinearSeries ℂ ℂ ℂ)
+    (U : Set ℂ) (z : ℂ) (hU : U ∈ nhds z) :
+  HasFPowerSeriesWithinAt f p U z ↔ HasFPowerSeriesAt f p z := by
+    simp only [HasFPowerSeriesWithinAt, HasFPowerSeriesAt]
+    refine ⟨fun H ↦ ?_, fun H ↦ ?_⟩
+    · have hzmem : z ∈ U := mem_of_mem_nhds hU
+      rw [Metric.mem_nhds_iff] at hU
+      obtain ⟨r', hr', hball⟩ := hU
+      let r'' : ENNReal := Option.some ⟨r', by linarith⟩
+      have hball' : EMetric.ball z r'' ⊆ U := by aesop
+      obtain ⟨renn, hr⟩ := H
+      use min (renn) r''
+      obtain ⟨r_le, r_pos, hs⟩ := hr
+      refine ⟨by aesop, by aesop, ?_⟩
+      · intros y a
+        apply hs
+        · have shift_ball (x y z : ℂ) (renn : ENNReal) :
+            x ∈ EMetric.ball y renn → z + x ∈ EMetric.ball (z + y) renn := by
+              simp only [EMetric.mem_ball, edist_add_left, imp_self]
+          have : z + y ∈ EMetric.ball (z + 0) (min renn r'') := by
+            apply shift_ball
+            exact a
+          have : z + y ∈ EMetric.ball z (min renn r'') := by aesop
+          have : z + y ∈ EMetric.ball z r'' := by aesop
+          have : z + y ∈ U := by aesop
+          aesop
+        · aesop
+    · obtain ⟨renn,hr⟩:= H
+      use renn
+      exact HasFPowerSeriesOnBall.hasFPowerSeriesWithinOnBall hr
```

Comment:
This entire block can go inside the `refine` above, it looks simple enough.
You're also not paying attention to spacing, the `obtain` statement is missing two spaces.

```lean
    refine ⟨fun H ↦ ?_, fun ⟨renn, hr⟩ ↦ ⟨renn, HasFPowerSeriesOnBall.hasFPowerSeriesWithinOnBall hr⟩⟩
```
This also saves you the bullet points, which helps reading the proof.
Since this doesn't go into one line, I'd move the `fun H ↦ ` part back to an `intro H`, but you can also make the `refine` span over 2 lines if you prefer it